### PR TITLE
DAG change detection

### LINKED ISSUE #4273: refactor(routing): consider VC-centric change detection model
The current hot-reload change detection uses a per-concern model: each
`ChangeDetector` independently scans all VCs for changes to the thing it
cares about (filters, routers, cluster definitions). Results are merged
by set-union.

This creates a gap when dependencies cross concern boundaries.
`ClusterDefinitionChangeDetector` documents one such gap explicitly:

> Walking the router DAG to find cluster definitions referenced
> transitively through routers is deferred.

The root cause isn't missing graph-walk code — `RouterChangeDetector`
already walks the router DAG. The issue is that it only looks for changed
*routers* at each node, not changed *cluster definitions* at the leaves.
No single detector owns the full transitive dependency graph of a VC.

A VC-centric model — "for each VC, walk everything it depends on and ask
whether any of it changed" — would close this class of gap structurally
rather than requiring each new concern to re-walk the graph. The existing
detectors wouldn't go away; they'd be rescoped as predicate services
(e.g. `clusterDefinitionDetector.hasChanged(name)`) that the VC walk
queries at each node, rather than independently scanning all VCs
themselves.

This is not urgent: the current gap only affects VCs that reach a
cluster definition exclusively through a router, and the router itself
hasn't changed. But it's worth considering before more concern types
are added.

Relates-to: #4242

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

Introduce a `RoutingGraphVisitor` / `RoutingGraphWalker` pattern for traversing a virtual cluster's routing graph (VC → Router DAG → ClusterDefinitions). The walker assumes validated input -- visitor parameters are non-null, and unresolvable names are skipped silently. `RouterGraphValidator.detectCycles` uses the walker; `validateReferences` remains a direct loop since it only needs one-hop checks.

Fix #4273 by merging the per-concern `RouterChangeDetector` and `ClusterDefinitionChangeDetector` into a single `RoutingGraphChangeDetector` that walks each VC's graph once with both predicates active.

Fixes #4273.
Follow-up to #4318.

### CODE REVIEW

**Comment by tombentley** [comment]
@robobario grateful if you could review this one

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/RouterGraphValidator.java [hunk 1]
```diff
@@ -84,55 +82,26 @@ private static void validateRouteIds(List<RouterDefinition> routerDefinitions) {
...
-                    }
-                    if (!visited.contains(target) && hasCycle(target, routersByName, visited, inStack, path)) {
-                        return true;
-                    }
-                }
+        @Override
+        public boolean enterRouter(RouterDefinition rd, WalkContext ctx) {
+            if (!ctx.isFirstVisit()) {
+                throw new IllegalConfigurationException(
+                        "Router definitions contain a cycle: " + String.join(" -> ", ctx.path()));
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/RouterGraphValidator.java:97]: just noting that this exception now contains the full path, rather than just the cycle which we had prior. I don't think that's a problem really.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/RoutingGraphChangeDetector.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+     * Returns {@code true} if any node reachable from {@code vc} in its routing graph
+     * satisfies either predicate.
+     *
+     * @param vc                   the virtual cluster whose graph is walked
+     * @param routersByName        all router definitions indexed by name
+     * @param clustersByName       all cluster definitions indexed by name
+     * @param routerNamePredicate  test applied to each visited router definition's name
+     * @param clusterNamePredicate test applied to each cluster definition's name at a leaf
+     */
+    public static boolean anyInClusterGraph(VirtualCluster vc,
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/RoutingGraphChangeDetector.java]: the class is package private so we should drop the visibility as much as we can

**Review by robobario — APPROVED** [review]
LGTM, I've put up a [PR](https://github.com/tombentley/kproxy/pull/65) with some further unit tests of the Walker for completeness. That includes a couple of edge cases and:

```
    1. When a child router refuses (enterRouter returns false), the parent's
       route loop breaks immediately and sibling routers are not visited.
    
    2. In a diamond topology (r1→r2→r4, r1→r3→r4), the shared router r4
       is visited exactly once; the cross-edge from r3 is silently skipped
       once r4 has been fully explored.
```
